stdlib/hashtbl.ml

```diff
@@ -109,74 +137,80 @@ let key_index h key =
 
 let add h key info =
   let i = key_index h key in
-  let bucket = Cons(key, info, h.data.(i)) in
+  let bucket = Cons{key; data=info; rest=h.data.(i)} in
   h.data.(i) <- bucket;
   h.size <- h.size + 1;
   if h.size > Array.length h.data lsl 1 then resize key_index h
 
+let rec remove_bucket h i key prec = function
+  | Empty ->
+      ()
+  | (Cons {key=k; rest=next}) as c ->
+      if compare k key = 0
+      then begin
+        h.size <- h.size - 1;
+        match prec with
+        | Empty -> h.data.(i) <- next
+        | Cons c -> c.rest <- next
+      end
+      else remove_bucket h i key c next
```

Comment:
For basic datastructures, I prefer to get a good understanding of what the resulting code looks like.  Anyway, I'm also interested in performances with js_of_ocaml; and flambda hasn't landed yet.